### ISSUE TITLE
Wrong type inference when using refine function

### ISSUE THREAD

**Issue by balazsszekely** [colinhacks/zod#638]
Hi, first of all thanks for this great library.

Unfortunately something went wrong after version 3.1.0 with the refine function.

Here is the code snippet:
```
import { z } from "zod";

const schema = z.object({
  date: z.date().refine((d) => true, "Invalid.")
});

type t = z.infer<typeof schema>;

const test: t = {}; //should be invalid
const test2: t = { date: "" }; //should be invalid

const test3: t = { date: new Date() }; //OK
```
Using version 3.1.0 the inferred type is ok:
```
type t = { date: Date }
```
but using later versions (3.2.0-3.8.2) the inferred type is:

```
type t = { date?: any }
```

https://codesandbox.io/s/zod-refine-issue-v203y?file=/src/index.ts

Can you fix it?
Is there any workaround for this bug?

Thanks,
Balazs

**Comment by balazsszekely** [comment]
Ok, I found what was the problem.

I saw in releases tab:

v3.2: Fix TypeScript 4.3 recursive type issues

So I gave it a try to upgrade typescript to v4.4.2, and it WORKS! :)

It would be great if you could update this section: https://github.com/colinhacks/zod#typescript-requirements .

Thanks

**Comment by colinhacks** [comment]
This should be fixed for TS 4.1-4.3 in Zod 3.9.
